naga/src/front/atomic_upgrade.rs

```diff
@@ -0,0 +1,373 @@
+
+        let constant = self.module.constants.try_get(handle)?.clone();
+        padding.debug("constant: ", &constant);
+
+        let new_constant = Constant {
+            name: constant.name.clone(),
+            ty: self.upgrade_type(constant.ty)?,
+            init: self.upgrade_expression(None, constant.init)?,
+        };
+
+        if constant != new_constant {
+            padding.debug("inserting new constant: ", &new_constant);
+            let new_handle = self
+                .module
+                .constants
+                .append(new_constant, self.module.constants.get_span(handle));
+            padding.debug("new constant handle: ", &new_handle);
+            Ok(new_handle)
+        } else {
+            Ok(handle)
+        }
+    }
+
+    fn upgrade_override(&mut self, handle: Handle<Override>) -> Result<Handle<Override>, Error> {
+        let padding = self.inc_padding();
+        padding.debug("upgrading override: ", handle);
+
+        let o = self.module.overrides.try_get(handle)?.clone();
+        padding.debug("override: ", &o);
+
+        let new_o = Override {
+            name: o.name.clone(),
+            id: o.id,
+            ty: self.upgrade_type(o.ty)?,
+            init: self.upgrade_opt_expression(None, o.init)?,
+        };
+
+        if o != new_o {
+            padding.debug("inserting new override: ", &new_o);
+            let new_handle = self
+                .module
+                .overrides
+                .append(new_o, self.module.overrides.get_span(handle));
+            padding.debug("new override handle: ", &new_handle);
+            Ok(new_handle)
+        } else {
+            Ok(handle)
+        }
+    }
+
+    fn upgrade_expression(
+        &mut self,
+        maybe_fn_handle: Option<Handle<Function>>,
+        handle: Handle<Expression>,
+    ) -> Result<Handle<Expression>, Error> {
+        let padding = self.inc_padding();
+        padding.debug("upgrading expr: ", handle);
+        let expr = if let Some(fh) = maybe_fn_handle {
+            let function = self.module.functions.try_get(fh)?;
+            function.expressions.try_get(handle)?.clone()
+        } else {
+            self.module.global_expressions.try_get(handle)?.clone()
+        };
+
+        padding.debug("expr: ", &expr);
+        let new_expr = match expr.clone() {
+            Expression::AccessIndex { base, index } => Expression::AccessIndex {
+                base: self.upgrade_expression(maybe_fn_handle, base)?,
+                index,
+            },
+            Expression::GlobalVariable(var) => {
+                Expression::GlobalVariable(self.upgrade_global_variable(var)?)
+            }
+            Expression::LocalVariable(var) => Expression::LocalVariable(
+                self.upgrade_local_variable(maybe_fn_handle.ok_or(Error::NoFunction)?, var)?,
+            ),
+            _ => {
+                return Err(Error::Unsupported);
+            }
+        };
```

Comment:
I'm purposefully only matching on the bare minimum here to make the diff as small as possible. I actually have the full match in a branch, but this way it's easier to see the structure of the PR.